### PR TITLE
Added real Log Rotation for jobs

### PR DESCRIPTION
**Pull request by awittha**
Sales Pitch
=====

A new section on the "Manage > Configure" page:

log rotation sales pitch

Context
=====

Since time immemorial, Jenkins masters have been running out of disk space and folks have been looking for ways to clean up old logs. Just about every search's answer is the [LogRotator](https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/tasks/LogRotator.java). Even if folks don't read the source code, their attempts to use the LogRotator quickly reveal:

https://github.com/jenkinsci/jenkins/blob/89e25fce5e5d8dbe65600ccff13950a1ad232d23/core/src/main/java/hudson/tasks/LogRotator.java#L55

The _distilled_ problem statement is that the only built-in tool for managing retention of logs created by user actions in Jenkins is _called_ a LogRotator but doesn't meet folks' expectations of how Log Rotation works.

The first major disconnect is that LogRotator has two axes for deciding whether to rotate "old" logs out - quantity and age - but only checks to see if logs should be rotated in response to changes of one of those axes - quantity.

That is to say, when you run a new build (and change the total quantity of builds), the LogRotator checks to see if you've got too many builds. But when you just _wait_ and your builds get _old_, the LogRotator is never engaged and fails to clean up old builds.

The second major disconnect is that LogRotators must be added individually and by hand (or with forethought, by some kind of configuration sharing mechanism e.g. configuration slicing or shared libraries) to every job whose logs should rotate. There is no way to "set up Log Rotation on a Jenkins Master" - it must be done to every project.

Combined with the first issue of "logs don't rotate until a build runs," this makes it very difficult to use the LogRotator to get a Jenkins master out of a bad state. The existing LogRotator is most useful if it is part of every job from the beginning of the master's life. That... is often not the case.

And heaven help you if you need to _update_ the log retention rules across a Jenkins master; you must update _and run_ every project on the master to do this!

We can change all that!

Contents
=====

This Pull Request adds a section to the "Manage > Configure" panel of Jenkins. By default, nothing happens, matching existing behavior.

default log rotation

When enabled, periodic and master-wide log rotation options become available:

log rotation options

This allows administrators to configure periodic, master-wide Log Rotation policies. They can configure different policies for jobs that define a LogRotator and jobs that do _not_ define one.

Now,

1. Log Rotation rules will run periodically, allowing jobs with _old_ builds to be cleaned up on schedule instead of just jobs with _too many_ builds.
2. The ability to rotate logs for builds that didn't define a Log Rotator by hand makes it easy to use this to recover a master from a bad (disk full, many jobs with no rotator) state.
3. The ability to ignore jobs' own Log Rotators in favor of global rules allows administrators to enforce a master-wide maximum log retention policy.

Details / Notes
=====

1. This adds a `GlobalConfiguration` section that configures an `AsyncPeriodicWork` extension.
2. I have included a Spanish translation to validate that the changes' text is properly abstracted & ready for localization.
3. I have added the ability to control `java.util.logging` verbosity during unit tests with Maven profiles.

Test Verbosity
-----

Unit tests that pass through `LOGGER.*` aren't chatty anymore.

Run `mvn test -Pverbose-test-logging` to get more `INFO` (info-level), or `mvn test -Pvery-verbose-test-logging` to show _all_ log messages encountered during tests.

PR Template
=====

### Proposed changelog entries

* Added ability to configure global log rotation policy

### Submitter checklist

- [ ] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [ ] ~For dependency updates: links to external changelogs and, if possible, full diffs~ (N/A)

### Desired reviewers

(none in particular)

Questions
=====

1. Should I create a JIRA issue for this change?

### CODE REVIEW

**Comment by daniel-beck** [comment]
> Even if folks don't read the source code, their attempts to use the LogRotator quickly reveal:

😁 Clever, but I think the actual disconnect between name and behavior that the comment hint at is indicated by the name of its replacement: _build discarder_.

**Comment by awittha** [comment]
Oh! I just realized that the MIT license block I grabbed out of the header of some random file (thinking it was the same block in ever file) actually had specific copyrights in it.

**Comment by awittha** [comment]
May I omit a specific copyright line in the header of the MIT license altogether, or will that cause problems?

File: core/src/main/java/hudson/tasks/LogRotator.java [hunk 1]
```diff
@@ -297,6 +299,16 @@ private String toString(Integer i) {
         if (i==null || i==-1)   return "";
         return String.valueOf(i);
     }
+    
+    @Override
+    public String toString() {
+    	return Objects.toStringHelper( this )
```
Comment (awittha) [inline core/src/main/java/hudson/tasks/LogRotator.java]: Added a human-readable ToString to LogRotator. When a globally-defined LogRotator is applied to a project (conceivably resulting in the _deletion_ of build records), we emit a very complete message to the Jenkins log recorder, which includes the LogRotator used. This toString makes it easier for that printout to show exactly which rules were applied.

**Review by res0nance — CHANGES_REQUESTED** [review]
On your sales pitch I'm 100% sold on this feature, I'd like to know how others feel

**Comment by daniel-beck** [comment]
Regarding the overall feature and sales pitch, my **first impression**:

I really like the idea of a global option that limits how many builds are kept around in jobs without dedicated configured build deletion. In larger orgs (more than a single, flat team?) it's wasting a lot of time to keep disk usage low enough so the Jenkins master disk can handle it. 👍👍 for adding such a feature into core.

However, the chosen implementation seems to be pretty complex. Anything in core (and to some extent in plugins that are installed by default) gets exposed to every newcomer admin. A list of independent retention rules with regular expressions matching job names (beware of renames!), … it seems a bit much. This is the kind of depth I'd expect us to deliver in an optional plugin, so that admins that need this flexibility can have it, while not making the out-of-the-box experience more difficult than it needs to be. Clearly you're a very experienced admin, and it seems you're writing a feature suitable for other very experienced admins.

The description clearly identifies the limits of when existing "log rotation" runs, and how configuration changes impact it, but for that specifically, a new `AsyncPeriodicWork` iterating over all jobs and calling `#logRotate()` would do it -- a System Groovy script doing something similar is how I got by at `$previousJob`.

Not having tried this manually, and after only limited code reading, I would guess that it is currently not possible to mark specific builds to be kept forever in jobs with no dedicated build discarder. That should be possible, otherwise this will end up deleting important information on instances that do not only use Jenkins for CI and have a notion of "important builds".

So, IMO, there's a lot to like here:

- a global fallback retention policy applying to jobs without a more specific policy makes sense. 
- Running log rotation periodically so old age is adequately responded to is long overdue.

This seems to be missing:

- Allow marking builds to be kept forever and that takes precedence even over the global policy

However, most of the options seem like a bit much for a built-in feature. Something like the following seems to address all of the stated problems (with newly periodic runs of existing build discarders simply existing without UI):

```
[x] Discard old builds of jobs that have no specific configuration to do so

[30] days to keep builds
[50] number of builds to keep
[Advanced…]
     [  ] days to keep artifacts
     [  ] number of builds with artifacts to keep
```

**Comment by awittha** [comment]
(force-push fixed license copyright text)

**Comment by awittha** [comment]
@daniel-beck 

Critical Build Retention
====

> I would guess that it is currently not possible to mark specific builds to be kept forever in jobs with no dedicated build discarder.

LogRotator takes care of this. Since the log rotation for builds that don't define their own rules is done by `LogRotator` itself, all of the [built-in criteria for retaining important builds](https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/tasks/LogRotator.java#L224-L242) are preserved and in effect. The "global policy" isn't anything original or out-of-band, it's just providing a `LogRotator` instance for builds that didn't define their own `LogRotator` instance.

I've personally witnessed that (via the Jenkins log recorder) on my own master.

Complexity of Options
=====

> However, the chosen implementation seems to be pretty complex.

My personal use-case that drove the development of this needed this level of capability. Originally I intended to build this all in a Plugin, but since all of the affected classes were part of `jenkins-core` and this seemed like a problem that _everyone_ had, I felt it would be better to contribute it to core.

The thing that _I_ do with the regular expressions is to give `.*/master` a longer retention policy than `.*`, so that jobs' primary branches keep more history, for longer, than PR and feature branches.

This is good feedback, as I wasn't sure how to default the global rules intuitively.

> newly periodic runs of existing build discarders simply existing without UI):

I would've liked that, too, but as I'm not a Jenkins product owner, I wasn't comfortable making the decision and assertion that everyone should get _new_ default behavior. If you are, then I'd happily change the default.

> ```
> [x] Discard old builds of jobs that have no specific configuration to do so
> 
> [30] days to keep builds
> [50] number of builds to keep
> [Advanced…]
>      [  ] days to keep artifacts
>      [  ] number of builds with artifacts to keep
> ```

However, the ability to override & configure the behavior of runs _with_ existing rotators is borne out of first-hand experience:

I had a master where every job had a LogRotator defined from the beginning of the master, but it had a very lenient policy (1000 builds and 90 days). When the master eventually filled up its disk... I was "stuck" having to update and run every single build on every single branch of every single project, to apply a newer, tighter policy. This was not tenable at the scale of the master at the time it ran out of disk space.

Being able to set the global rules in effect _even for builds that define their own rotator_ means that 

1. nightly, I'm guaranteed that everything will rotate with the policy I want
2. no build can define a LogRotator with an egregiously large retention policy and thwart my disk management attempts

There are other masters at my company that are administered under a different semantic, where their admins would probably always defer to each job's own rotator, and just use the "run it nightly" capability.
 
In summary, I'm interested in ideas regarding more-intuitive default behavior (especially for global rules [default regex = .* instead of blank? Have a tooltip? etc]) _and_  clearer wording / default values, but I stand by the functionality as it stands as the minimum. At least, the minimum that _I_ need.

I've also deliberately _not_ written any `help-XXX.html` files, hoping that the UI wouldn't need them and waiting on feedback to identify things that were clear to me but not to others (and thereby, worthy of inclusion in a dedicated tooltip).

File: core/src/main/java/hudson/tasks/LogRotatorPeriodicTask.java [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+                                                        
+                            BuildDiscarder discarder = job.getBuildDiscarder();
+                                                        
+                            try {
+                            
+                                if( discarder instanceof LogRotator ) {
+                                    // the job has defined its own LogRotator                                    
+                                    switch( config.getPolicyForJobsWithCustomLogRotator() ) {
+                                    case NONE:
+                                        LOGGER.log( FINER, Messages.LogRotatorPeriodicTask_hasRotator_policyNONE( job.getFullName() ) );
```
[the remark this answers is not in the document]
Comment (awittha) [reply core/src/main/java/hudson/tasks/LogRotatorPeriodicTask.java]: Why? I haven't seen this before, and it seems to work without it.

File: core/src/main/java/hudson/tasks/LogRotatorPeriodicTask.java [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        
+        LogRotatorConfiguration config = GlobalConfiguration.all().get( LogRotatorConfiguration.class );
+        
+        if( config != null ) {
+            // there is config
+            if( config.isEnableRotation() && config.getUpdateIntervalHours() > 0 ) {
+                // and we should update periodically
+                
+                long millis_since_last_update = System.currentTimeMillis() - config.getLastRotated();
+                double hours_since_last_update = millis_since_last_update / 1000.0 / 60.0 / 60.0;
```
[the remark this answers is not in the document]
Comment (awittha) [reply core/src/main/java/hudson/tasks/LogRotatorPeriodicTask.java]: I generally prefer keeping inline time math as just arithmetic. If preferring TimeUnit for internal logic as well as user-interface code is a pattern in jenkins-core, I'll happily match that.

File: core/src/main/java/hudson/tasks/LogRotatorPeriodicTask.java [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+        }
+    }
+    
+    /*
+     * Non-javadoc: will run every minute, but only to check if sufficient time has passed to execute.
+     * Since AsyncPeriodicWork's timer is non-configurable, we need to run every minute, and check
+     * if there's actually anything to do inside the execute(...) 
+     */
+    @Override
+    public long getRecurrencePeriod() { return MIN; }
```
[the remark this answers is not in the document]
Comment (awittha) [reply core/src/main/java/hudson/tasks/LogRotatorPeriodicTask.java]: That was one of my uncertainties. In the interest of meeting folks' expectations as completely and consistently as possible, `MIN` is a good choice: the _most_ that a log rotation period (defined in hours) could "miss" the exact mark is 59s, or 1/60th (ish) of the rotation period.

However, if there's tolerance for some imprecision, I would say that this should _absolutely_ be HOUR.

You can imagine the situation in which someone saves a configuration to rotate logs hourly 1 minute since the last check. 59 minutes later, it checks again but 1 hour has not passed, and then an hour later it checks and does execute... and the user perceives a 2 hour wait, nearly a 100% difference from the expected period.

I believe HOUR is _fine_ here, but I chose to err on the side of precision for the initial PR.

File: test/src/test/java/hudson/tasks/LogRotatorPeriodicTaskTest.java [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+        
+        // set global Log Rotation policy for jobs with a custom rotator:
+        // do nothing
+        LogRotatorConfiguration config = GlobalConfiguration.all().get( LogRotatorConfiguration.class );
+        assertNotNull( config );
+        config.setPolicyForJobsWithCustomLogRotator( LogRotationPolicy.NONE );
+        
+        // this global LogRotator should match all jobs.
+        // but since the policy is "do nothing," it should not be used.
+        config.setGlobalLogRotators(Arrays.asList( new LogRotatorMapping[] {
```
[the remark this answers is not in the document]
Comment (awittha) [reply test/src/test/java/hudson/tasks/LogRotatorPeriodicTaskTest.java]: I prefer to be explicit; once in a blue moon the compiler is happier with the generic outcome when it's all spelled out unambiguously. It's just a few extra keypresses in an IDE with autocomplete, but it may prevent a pernicious headache.

In this case it seems to be unnecessary; I've updated the test class to omit the array.

File: core/src/main/java/hudson/tasks/LogRotatorPeriodicTask.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (awittha) [reply core/src/main/java/hudson/tasks/LogRotatorPeriodicTask.java]: Oh! It's to defer processing of the message until it actually has to get rendered in a log, speeding up the pass through the actual code, isn't it?

**Comment by awittha** [comment]
Idea to possibly simplify the "global rules:"

A further radio button toggle for global rules, "basic / advanced."

The "basic" view is the default, and shows just a single LogRotator's fields, and they're already filled in with a reasonable default retention.

The "advanced" view is the multiple-rotator, regex-driven scheme currently pictured.

Then, an admin that ticks the "enable log rotation" box doesn't need to do anything else except save the page, and every log on the master is rotating sensibly. But if they need more, they can dig into the "advanced" section, as the Jenkins UI has already conditioned them to do.

**Comment by daniel-beck** [comment]
> ## Critical Build Retention

There is no UI to set the field that gets checked by https://github.com/jenkinsci/jenkins/blob/4f0436abb1ef55372152e13f4b6953f1f216d1ba/core/src/main/java/hudson/tasks/LogRotator.java#L225 while there is no log rotator configured on the job: https://github.com/jenkinsci/jenkins/blob/4f0436abb1ef55372152e13f4b6953f1f216d1ba/core/src/main/resources/hudson/model/Run/logKeep.jelly#L30

So any job relying on the global fallback build discarder would be unable to define builds to retain forever.

> ## Complexity of Options

It should be fairly straightforward to define an extension point globally, with a simple default implementation shipped with core, and a plugin extending it with the regexes etc.

The UI would probably look not much different than what you'd get with a built-in basic/advanced toggle, probably similar to the `MarkupFormatter` config UI.

If that sounds good, I can try to sketch something up.

**Comment by daniel-beck** [comment]
Oh, and ideally this change would also look critically at the existing UI for discarding old builds on a per-job basis: It's no longer _Discard old builds_ but rather _Override global build discarder setting_ (at least as long as there's no global override -- another reason I'm not a big fan of that, at least in core).

**Comment by awittha** [comment]
> So any job relying on the global fallback build discarder would be unable to flag builds to be retained forever.

Yes - but the critical ones (lastSuccess, etc) will be retained. I think it's thoroughly reasonable that a job won't be able to mark builds as "keepers forever," without defining its own retention rules!

**Comment by awittha** [comment]
> It should be fairly straightforward to define an extension point globally, with a simple default implementation shipped with core, and a plugin extending it with the regexes etc.
> 
> The UI would probably look not much different than what you'd get with a built-in basic/advanced toggle, probably similar to the MarkupFormatter config UI.
> 
> If that sounds good, I can try to sketch something up.

Sounds fine to me; please do!

I haven't built in-plugin extension points before... and I also have a different process that I must follow to contribute to existing OSS repos vs creating new ones - so I cannot create &  publish a new separate plugin that adds the "advanced" behavior (at least not in anything that resembles a timely fashion).

**Comment by awittha** [comment]
> Oh, and ideally this change would also look critically at the existing UI for discarding old builds on a per-job basis: It's no longer Discard old builds but rather Override global build discarder setting (at least as long as there's no global override -- another reason I'm not a big fan of that, at least in core).

I think it's fair to leave the existing text alone; it's still accurate _unless_ the administrator chooses to override it.

Since that's a configurable toggle, it would be easy to surface a note or warning, or change the UI text, if a global override were in place. But, I'm not sure that's really necessary.

**Comment by daniel-beck** [comment]
> Trivial to add a toggle to expose the `isKeep` ui if no local build discarder _but also_ there's global log rotation, right?

Yes, it's not a big challenge to implement, but it seems necessary to me to make this global fallback build discarder feature work well, and is currently missing.

> Sounds fine to me; please do!

ETA… uh… next two weeks, probably? Depends on what else I fill my weekends with.

**Comment by daniel-beck** [comment]
CC/FYI @rtyler who I expect will welcome this feature with open arms, given what some of our runbooks do 😁

**Comment by daniel-beck** [comment]
On-holding this for a bit based on recent comments.

@awittha I poked you in Gitter, PTAL.

File: core/src/main/java/hudson/tasks/LogRotatorPeriodicTask.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (res0nance) [reply core/src/main/java/hudson/tasks/LogRotatorPeriodicTask.java]: I think TimeUnit is preferable and I often do see most devs defaulting to that. I feel TimeUnit is just easier to tell durations from a glance.

File: core/src/main/java/hudson/tasks/LogRotatorPeriodicTask.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (res0nance) [reply core/src/main/java/hudson/tasks/LogRotatorPeriodicTask.java]: Yes, for any processing to build your logging string this defers its creation till its needed. In the past a user would have to construct the whole string before calling LOGGER.log but this created the pattern where users would check outside the function leading to
```
if (LOGGER.isLoggable(FINER)) {
LOGGER.log()
}
```
Using a lambda just simplifies it since internally `log` checks if the logger is loggable at the specified level first.

File: core/src/main/java/hudson/tasks/LogRotatorPeriodicTask.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jvz) [reply core/src/main/java/hudson/tasks/LogRotatorPeriodicTask.java]: Nit: Would also be nice to maintain the Java code style of variable names using camelCase instead of Python style.

**Comment by rtyler** [comment]
omg let's ship it :tada:

**Comment by awittha** [comment]
@daniel-beck which gitter? I don't really ever use Gitter except when I know I'm going to go look for something in particular. I don't seem to have any notifications in my e-mail about unread gitter messages relating to this.

**Comment by timja** [comment]
> @daniel-beck which gitter? I don't really ever use Gitter except when I know I'm going to go look for something in particular. I don't seem to have any notifications in my e-mail about unread gitter messages relating to this.
There's only one gitter I think? but this is the url for the jenkins room, you should have a DM from @daniel-beck I assume
https://gitter.im/jenkinsci/jenkins

**Comment by timja** [comment]
(conflicts currently)

**Comment by daniel-beck** [comment]
This one slipped my mind, sorry about that. Planning to take care of it this weekend.

File: core/src/main/java/hudson/tasks/LogRotatorPeriodicTask.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (varyvol) [reply core/src/main/java/hudson/tasks/LogRotatorPeriodicTask.java]: I agree with both suggestions.

File: core/src/main/java/hudson/tasks/LogRotatorPeriodicTask.java [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+                    // actually do it
+                    LOGGER.log( getNormalLoggingLevel(), Messages.LogRotatorPeriodicTask_starting() );
+                    
+                    Jenkins jenkins = Jenkins.getInstanceOrNull();
+                    
+                    if( null == jenkins ) { return; }
+                    
+                    for( Item item : jenkins.getAllItems() ) {
+                        
+                        for( Job<?,?> job : item.getAllJobs() ) {
```
Comment (varyvol) [inline core/src/main/java/hudson/tasks/LogRotatorPeriodicTask.java]: I'm not sure about this, it could be a huge performance impact.

File: core/src/main/java/hudson/model/LogRotatorConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+import org.kohsuke.stapler.QueryParameter;
+
+import hudson.Extension;
+import hudson.tasks.LogRotator;
+import hudson.util.FormValidation;
+import jenkins.model.GlobalConfiguration;
+
+/**
+ * @author awitt
+ * @since 2.203
```
[the remark this answers is not in the document]
Comment (fcojfernandez) [reply core/src/main/java/hudson/model/LogRotatorConfiguration.java]:
```suggestion
 * @since TODO
```

File: core/src/main/java/hudson/tasks/LogRotatorPeriodicTask.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (varyvol) [reply core/src/main/java/hudson/tasks/LogRotatorPeriodicTask.java]: I also think 1 hour is fine. You described the worst scenario but the average one should not be that bad. Also log rotation seems to me like something to think about in long-term, so shouldn't be worrying if one execution is missed.

File: core/src/main/java/hudson/model/LogRotatorConfiguration.java [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+        }
+        
+        return FormValidation.ok();
+    }
+    
+    /**
+     * How to rotate the logs of a given job.
+     * 
+     * @author awitt
+     * @since 2.203
```
Comment (fcojfernandez) [inline core/src/main/java/hudson/model/LogRotatorConfiguration.java]:
```suggestion
     * @since TODO
```

**Comment by timja** [comment]
Are both this and https://github.com/jenkinsci/jenkins/pull/4368 needed  ?
@awittha and @daniel-beck  ?

**Comment by daniel-beck** [comment]
#4368 supersedes this proposal. Unless I drop the ball badly, I expect to finish that PR soon and then this one can be closed unmerged.

**Comment by daniel-beck** [comment]
I consider #4368 done, and propose that we close this one.

**Comment by timja** [comment]
@awittha any objections to closing this?

**Comment by awittha** [comment]
None. See y'all in #4368
